lib/types/trie.go

```diff
@@ -8,46 +8,69 @@ type trieNode struct {
 }
 
 func (t *trieNode) insert(s string) {
-	if len(s) == 0 {
-		t.isLeaf = true
-		return
-	}
+	runes := []rune(s)
 
-	// mask creation of the trie by initializing the root here
 	if t.children == nil {
-		t.children = make(map[rune]*trieNode)
+		t.children = map[rune]*trieNode{}
 	}
 
-	rStr := []rune(s) // need to iterate by runes for intl' names
-	last := len(rStr) - 1
-	if c, ok := t.children[rStr[last]]; ok {
-		c.insert(string(rStr[:last]))
-		return
+	ptr := t
+	for i := len(runes) - 1; i >= 0; i-- {
+		r := runes[i]
+		c, ok := ptr.children[r]
+
+		if !ok {
+			ptr.children[r] = &trieNode{children: map[rune]*trieNode{}}
+			c = ptr.children[r]
+		}
+
+		ptr = c
 	}
 
-	t.children[rStr[last]] = &trieNode{children: make(map[rune]*trieNode)}
-	t.children[rStr[last]].insert(string(rStr[:last]))
+	ptr.isLeaf = true
 }
 
-func (t *trieNode) contains(s string) (matchedPattern string, matchFound bool) {
-	s = strings.ToLower(s)
-	if len(s) == 0 {
-		if t.isLeaf {
-			return "", true
+func (t *trieNode) contains(s string) (string, bool) {
+	rs := []rune(strings.ToLower(s))
```

Comment:
You could conside to use `unicode.ToLower` directly in the `for` below

Reply:
Have you seen [c14be6b](https://github.com/grafana/k6/pull/2777/commits/c14be6b8ea70900d132f7ad87841c1424114fe6f)

> IMO trieNode should be generic enough because it's a DS, it should be usable in other part of the codebase. Currently we only use in HostnameTrie and AddresTrie (not in master a separate PR). Because hostname is not case sensitive we lowercased it in trieNode.contains() too, last commit fixes that.

Which way should we proceed?

Reply:
I missed it, thanks for pointing me there. I agree it is better to keep the data structure as generic as possible.